src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
+
+class QosDatastruct {
+  enum {
+    RGW_QOS_OPS,
+    RGW_QOS_CONCURR,
+    RGW_QOS_BW,
+  };
+  struct qos_entry {
+    // 3 is the enum size
+    std::atomic_int64_t read[3] = {0};
+    std::atomic_int64_t write[3] = {0};
+    int64_t ts = 0;
+    std::mutex ts_lock;
+  };
+  const size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::mutex insert_lock;
+  std::condition_variable *cv;
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
+    if((strcmp(method,"GET") == 0) || (strcmp(method,"HEAD") == 0))
+      return true;
+    return false;
+  }
+    // find or create an entry, and return its iterator
+  auto find_or_create(const std::string& key) {
+    insert_lock.lock();
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.emplace(std::piecewise_construct,
+                               std::forward_as_tuple(key),
+                               std::forward_as_tuple()).first;
+    insert_lock.unlock();
+    return ret;
+  }
+
+  bool increase_entry_read(hash_map::iterator& it, int64_t ops_limit, int64_t bw_limit) {
+    int64_t ops = it->second.read[RGW_QOS_OPS];
+    int64_t concur = ++it->second.read[RGW_QOS_CONCURR];
+    int64_t bw = it->second.read[RGW_QOS_BW];
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((ops - 1 < 0 ||
+      concur > ops_limit) && (ops_limit > 0)) ||
+      (bw < 0 && bw_limit > 0))
+      return true;
+    // we don't want to reduce ops' tokens if we've rejected it.
+    --it->second.read[RGW_QOS_OPS];
+    return false;
+  }
+  
+  bool increase_tokens(hash_map::iterator& it, int64_t curr_timestamp,
+                       RGWQoSInfo& tenant) 
+  {
+    it->second.ts_lock.lock();
+    int64_t read_ops = (tenant.max_read_ops * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t read_bw = (tenant.max_read_bytes * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t write_ops = (tenant.max_write_ops * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t write_bw = (tenant.max_write_bytes * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
```

Comment:
do we have a way to account for burst also? what would that look like?

my understanding is that we're refilling the buckets every second, meaning that we don't remember anything past 1 second. this really doesn't seem right to me, when it's common for requests to span several seconds or even minutes

i think it would be much more natural to think about these limits as rates per minute rather than second. then we could use a pure token bucket algorithm without relying on the CONCURR tracking piece. what do you think?

Reply:
> Because RGW will not drop running request even if it reached the bandwidth limit, I added a debt of maximum of 60 seconds, so the tenant will not be able to send more requests in maximum of 60 seconds (maybe we should make it configurable, what do you think?)

would tracking the rates per-minute also address this concern?